### PR TITLE
feat(bun/test): Implement "bail" option for "bun test"

### LINKED ISSUE #3044: bun:test - option to abort the whole test run if one test fails
### What is the problem this feature would solve?

Exit the test suite immediately upon N number of failing test suite

```ts
test('0', () => {
  expect(true).toBeTruthy()
})
test('1', () => {
  expect(false).toBeTruthy() // this fails
})
test('2', () => {
  // this test still runs even the previous one failed
})
```

### What is the feature you are proposing to solve the problem?

```ts

// new option (like @playwright/test)
test.configure({ maxFailures: 1 })

// or like jest: https://jestjs.io/docs/cli#--bailn

test('0', () => {
  expect(true).toBeTruthy()
})
test('1', () => {
  expect(false).toBeTruthy() // this fails
})
test('2', () => {
  // this test should not run if the previous one fails
})
```

### What alternatives have you considered?

_No response_

### PR DESCRIPTION
**Pull request by v3nntrnx**
https://jestjs.io/docs/configuration#bail-number--boolean

Fixes #3044

### CODE REVIEW

File: src/cli/test_command.zig [hunk 1]
```diff
@@ -200,6 +201,14 @@ pub const CommandLineReporter = struct {
         this.summary.fail += 1;
         this.summary.expectations += expectations;
         this.jest.tests.items(.status)[id] = TestRunner.Test.Status.fail;
+
+        if (this.jest.bail == this.summary.fail) {
```
Comment (Electroid) [inline src/cli/test_command.zig:205]: We should reuse the same code for printing the existing summary here so we don't have 2 different formats. I like the "bail out" message, so we can keep that if it's a bail.

File: src/cli.zig [hunk 2]
```diff
@@ -383,7 +384,14 @@ pub const Arguments = struct {
                     };
                 }
             }
-            ctx.test_options.update_snapshots = args.flag("--update-snapshots");
+            if (args.option("--bail")) |bail| {
+                if (bail.len > 0) {
+                    ctx.test_options.bail = std.fmt.parseInt(u32, bail, 10) catch |e| {
+                        Output.prettyErrorln("--bail expects a number: {s}", .{@errorName(e)});
```
Comment (Electroid) [inline src/cli.zig:390]: Should add a check that `bail > 0`, since `bail = 0` doesn't make sense

Comment (TiranexDev) [reply src/cli.zig:390]: Oh yeah was a little typo

Comment (v3nntrnx) [reply src/cli.zig:390]: Wait, it checks for length of the parameter, so if it doesn't exist

Comment (TiranexDev) [reply src/cli.zig:390]: other numbers do it too

File: src/cli/test_command.zig [hunk 3]
```diff
@@ -81,6 +81,7 @@ pub const CommandLineReporter = struct {
     summary: Summary = Summary{},
     prev_file: u64 = 0,
     repeat_count: u32 = 1,
+    start_time: i128 = 0,
```
Comment (Jarred-Sumner) [inline src/cli/test_command.zig]: start_time shouldn't need to be stored here, it's a global var

File: src/cli.zig (hunk 2 above)
Comment (v3nntrnx) [reply src/cli.zig:390]: oh, i just realized by what you mean. sure i'll do it

File: src/cli.zig [hunk 4]
```diff
@@ -383,7 +384,19 @@ pub const Arguments = struct {
...
+                        Global.exit(1);
+                    };
+
+                    if (ctx.test_options.bail == 0) {
+                        Output.prettyErrorln("--bail expects a number greater than 0", .{});
+                        Global.exit(1);
+                    }
+                }
```
Comment (Electroid) [inline src/cli.zig:400]: missing an `else` case, when you do `bun test --bail`, that implies a bail of `1`. so this should set `ctx.test_options.bail = 1;`

File: src/cli/test_command.zig [hunk 5]
```diff
@@ -246,6 +252,19 @@ pub const CommandLineReporter = struct {
         this.summary.expectations += expectations;
         this.jest.tests.items(.status)[id] = TestRunner.Test.Status.todo;
     }
+
+    pub fn printSummary(this: *CommandLineReporter, summary: Summary, start_time: i128, total_tests: ?u32) void {
```
Comment (Electroid) [inline src/cli/test_command.zig]: Shouldn't this also print all the skipped/todo tests? also give the summary of 0 pass, 0 fail, etc.? That's what I would expect

**Review by Electroid — CHANGES_REQUESTED** [review]

File: src/cli/test_command.zig (hunk 5 above)
Comment (v3nntrnx) [reply src/cli/test_command.zig]: well the problem is the total_tests is not very total until all of the tests are run, so when it bails out we don't know amount of all tests

Comment (v3nntrnx) [reply src/cli/test_command.zig]: or did you mean the runned_tests variable?

Comment (v3nntrnx) [reply src/cli/test_command.zig]: oh i got you

Comment (Electroid) [inline src/cli/test_command.zig]: Since `start_time` is always `bun.start_time`, this doesn't need to be a function argument.

File: src/cli/test_command.zig [hunk 6]
```diff
@@ -80,7 +80,7 @@ pub const CommandLineReporter = struct {
     last_dot: u32 = 0,
     summary: Summary = Summary{},
     prev_file: u64 = 0,
-    repeat_count: u32 = 1,
+    repeat_count: u32 = 0,
```
Comment (Electroid) [inline src/cli/test_command.zig]: Is this a drive-by bug fix? Can we make sure it doesn't break `--rerun-each`?

Comment (v3nntrnx) [reply src/cli/test_command.zig]: Idk how that happened, I swear-

**Review by paperclover — APPROVED** [review]
looks good, gonna let tests run. some non-related test fails will happen

**Review by paperclover — APPROVED** [review]
yeah

**Review by Jarred-Sumner — APPROVED** [review]
